src/libs/swaps/basis/rBasisSwap.ts

```diff
@@ -0,0 +1,613 @@
+  return {
+    publicKey,
+    account,
+  };
+}
+
+export const unstakeBasisInstruction = (
+  amount: number | BN | string,
+  userAuthority: PublicKey,
+  userToken: PublicKey,
+  userRedeemable: PublicKey,
+): TransactionInstruction => {
+  const dataLayout = BufferLayout.struct([
+    Layout.uint64('amount'),
+  ]);
+
+  const data = Buffer.alloc(dataLayout.span);
+  dataLayout.encode(
+    {
+      amount: new BN(amount),
+    },
+    data,
+  );
+
+  // userAuthority    M[ ] S[x]
+  // userToken        M[x] S[ ]
+  // userRedeemable   M[x] S[ ]
+  // redeemableMint   M[x] S[ ]
+  // tokenMint        M[ ] S[ ]
+  // stakingAccount   M[ ] S[ ]
+  // tokenVault       M[x] S[ ]
+  // tokenProgram     M[ ] S[ ]
+
+  const keys = [
+    { pubkey: userAuthority, isSigner: true, isWritable: false },
+    { pubkey: userToken, isSigner: false, isWritable: true },
+    { pubkey: userRedeemable, isSigner: false, isWritable: true },
+    { pubkey: new PublicKey(MINT_RBASIS), isSigner: false, isWritable: true },
+    { pubkey: new PublicKey(MINT_BASIS), isSigner: false, isWritable: false },
+    { pubkey: new PublicKey(PROGRAM_BASIS_STAKING_INSTANCE), isSigner: false, isWritable: false },
+    { pubkey: new PublicKey(PROGRAM_BASIS_STAKING_VAULT), isSigner: false, isWritable: true },
+    { pubkey: TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
+  ];
+
+  return new TransactionInstruction({
+    keys,
+    programId: new PublicKey(PROGRAM_BASIS_STAKING),
+    data: Buffer.concat([sighash(SIGHASH_GLOBAL_NAMESPACE, StakingInstructionNames[StakingInstruction.unstake]), data]),
+  });
+};
+
+export const unstakeBasis = async (
+  connection: Connection,
+  payer: Account,
+) => {
+  const ixs: TransactionInstruction[] = [];
+
+  // get associated token account for Basis (or create if doens't exist)
+  const BasisAccount = await Token.getAssociatedTokenAddress(
+    ASSOCIATED_TOKEN_PROGRAM_ID,
+    TOKEN_PROGRAM_ID,
+    new PublicKey(MINT_BASIS),
+    payer.publicKey,
+  );
+  const BasisAccountInfo = await connection.getAccountInfo(BasisAccount);
+  if (!BasisAccountInfo) {
+    const createBasisAtaIx = Token.createAssociatedTokenAccountInstruction(
+      ASSOCIATED_TOKEN_PROGRAM_ID,
+      TOKEN_PROGRAM_ID,
+      new PublicKey(MINT_BASIS),
+      BasisAccount,
+      payer.publicKey,
+      payer.publicKey,
+    );
+    ixs.push(createBasisAtaIx);
+  }
+  // DEBUG
+  // console.log('BASIS ACCOUNT', BasisAccount.toBase58());
+
+  // get associated token account for rBasis
+  const rBasisAccount = await Token.getAssociatedTokenAddress(
+    ASSOCIATED_TOKEN_PROGRAM_ID,
+    TOKEN_PROGRAM_ID,
+    new PublicKey(MINT_RBASIS),
+    payer.publicKey,
+  );
+  const rBasisAccountInfo = await connection.getAccountInfo(rBasisAccount);
+  const rBasisTokenAccount = await getTokenAccount(connection, rBasisAccount);
+  if (!rBasisAccountInfo || !rBasisTokenAccount) throw new Error('this function requires active rBasis associated token account');
```

Comment:
instead of throwing an error, we can just return. Not all wallets will have an rbasis ata

Reply:
👍